packages/core/src/middleware/koa-log.ts

```diff
@@ -1,13 +1,40 @@
-import { LogPayload, LogPayloads, LogResult, LogType } from '@logto/schemas';
+import { BaseLogPayload, LogPayload, LogPayloads, LogResult, LogType } from '@logto/schemas';
 import { Optional } from '@silverhand/essentials';
 import deepmerge from 'deepmerge';
 import { MiddlewareType } from 'koa';
 import { nanoid } from 'nanoid';
 
 import { insertLog } from '@/queries/log';
 
-export type WithLogContext<ContextT> = ContextT & {
-  log: <T extends LogType>(type: T, payload: LogPayloads[T]) => void;
+type MergeLog = <T extends LogType>(type: T, payload: LogPayloads[T]) => void;
+
+export type WithLogContext<ContextT> = ContextT & { log: MergeLog };
+
+const initLog = (baseLogPayload?: Readonly<BaseLogPayload>) => {
+  /* eslint-disable @silverhand/fp/no-let */
+  let type: Optional<LogType>;
+  let payload: LogPayload;
+  /* eslint-enable @silverhand/fp/no-let */
+
+  /* eslint-disable @silverhand/fp/no-mutation */
+  const mergeLog: MergeLog = (logType, logPayload) => {
+    if (type !== logType) {
+      // Reset payload when type changes
+      payload = { result: LogResult.Success, ...baseLogPayload };
+    }
+
+    type = logType;
+    payload = deepmerge(payload, logPayload);
+  };
+
+  const mergeLogError = (error: unknown) => {
+    payload = deepmerge(payload, { result: LogResult.Error, error: String(error) });
```

Comment:
no-need to deep merge this?

Reply:
Cherry-picked your commits. Thanks for your suggestion.